Add support of Azure standard load balancer and public IP

feiskyer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This PR adds a new configure `loadBalancerSku` and `excludeMasterFromStandardLB` to support Azure standard load balancer and public IP.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
First step of #60485

**Special notes for your reviewer**:

This is the first step of #60485, and only VMAS is supported yet. VMSS support will be added in a following PR.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Azure cloud provider now supports standard SKU load balancer and public IP. To use it, set cloud provider config with
{
  "loadBalancerSku": "standard",
  "excludeMasterFromStandardLB": true,
}

If excludeMasterFromStandardLB is not set, it will be default to true, which means master nodes are excluded to the backend of standard LB.

Also note standard load balancer doesn't work with annotation `service.beta.kubernetes.io/azure-load-balancer-mode`. This is because all nodes (except master) are added as the LB backends.
```

feiskyer commented:
/sig azure

feiskyer commented:
/retest

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard.go:
```diff
@@ -623,6 +630,22 @@ func (as *availabilitySet) ensureHostInPool(serviceName string, nodeName types.N
+			matches := backendPoolIDRE.FindStringSubmatch(backendPool)
+			if len(matches) == 2 {
+				lbName := matches[1]
+				if strings.HasSuffix(lbName, InternalLoadBalancerNameSuffix) == isInternal {
```
It is possible to pull master nodes into the pool, if it only has public/internal LB.

feiskyer replied:
Added a check before calling ensureHostInPool

feiskyer commented:
@karataliu As discussed offline, added two configs now. PTAL

```json
{
  "loadBalancerSku": "standard",
  "excludeMasterFromStandardLB": true,
}
```

If excludeMasterFromStandardLB is not set, master nodes are also added to the backend of standard LB. This is useful for small clusters, e.g. 3 nodes with HA, in which case master nodes are also running workload Pods.

karataliu commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -109,6 +112,12 @@ type Config struct {
+	LoadBalancerSku string `json:"loadBalancerSku" yaml:"loadBalancerSku"`
+	// ExcludeMasterFromStandardLB excludes master nodes from standard load balancer.
+	ExcludeMasterFromStandardLB bool `json:"excludeMasterFromStandardLB" yaml:"excludeMasterFromStandardLB"`
+
```
Should assign default values to those 2 somewhere. ExcludeMasterFromStandardLB should default to true?

feiskyer replied:
ExcludeMasterFromStandardLB  is bool, so it is default to false. This works for all deployments

karataliu replied:
Consider using pointers thus it'll be easy to change default value

ref: https://github.com/Azure/acs-engine/pull/1584#pullrequestreview-77863462

karataliu commented on pkg/cloudprovider/providers/azure/azure_wrap.go:
```diff
@@ -242,3 +243,11 @@ func (az *Cloud) newRouteTableCache() (*timedCache, error) {
 }
+
+func (az *Cloud) useStandardLoadBalancer() bool {
+	if strings.ToLower(az.LoadBalancerSku) == loadBalancerSkuStandard {
```
can return strings.EqualFold(az.LoadBalancerSku,loadBalancerSkuStandard) directly

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard.go:
```diff
@@ -123,7 +130,7 @@ func (az *Cloud) mapLoadBalancerNameToVMSet(lbName string, clusterName string) (
 func (az *Cloud) getLoadBalancerName(clusterName string, vmSetName string, isInternal bool) string {
 	lbNamePrefix := vmSetName
-	if strings.EqualFold(vmSetName, az.vmSet.GetPrimaryVMSetName()) {
+	if strings.EqualFold(vmSetName, az.vmSet.GetPrimaryVMSetName()) || az.useStandardLoadBalancer() {
```
Anywhere to add hint that selectLoadBalancer does not work for StandardLoadBalancer?

feiskyer replied:
Good catch, will add this check in getLoadBalancer()

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard_test.go:
```diff
@@ -157,10 +158,120 @@ func TestMapLoadBalancerNameToVMSet(t *testing.T) {
+		},
+		{
+			description:   "non-default standard external LB should map to empty string",
+			lbName:        "azuretest-internal",
```
external LB ?

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard_test.go:
```diff
@@ -157,10 +158,120 @@ func TestMapLoadBalancerNameToVMSet(t *testing.T) {
 	for _, c := range cases {
+		if c.useStandardLB {
+			az.Config.LoadBalancerSku = loadBalancerSkuStandard
+		}
```
should set LoadBalancerSku back for !useStandardLB. This works now since the Standard SKU cases are in the end.

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard_test.go:
```diff
@@ -157,10 +158,120 @@ func TestMapLoadBalancerNameToVMSet(t *testing.T) {
+	for _, c := range cases {
+		if c.useStandardLB {
+			az.Config.LoadBalancerSku = loadBalancerSkuStandard
+		}
```
Also, switch back for !useStandardLB

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard_test.go:
```diff
@@ -157,10 +158,120 @@ func TestMapLoadBalancerNameToVMSet(t *testing.T) {
+		if c.useStandardLB {
+			az.Config.LoadBalancerSku = loadBalancerSkuStandard
+		}
+		vmset := az.getLoadBalancerName(c.clusterName, c.vmSet, c.isInternal)
```
better 'loadbalancerName := '

feiskyer commented:
@karataliu addressed comments. PTAL

feiskyer commented:
/retest

karataliu commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -273,6 +284,11 @@ func (az *Cloud) selectLoadBalancer(clusterName string, service *v1.Service, exi
 				Location:                     &az.Location,
 				LoadBalancerPropertiesFormat: &network.LoadBalancerPropertiesFormat{},
 			}
+			if az.useStandardLoadBalancer() {
```
This will not be true now due to ' wantLb && !az.useStandardLoadBalancer()'

feiskyer replied:
ack

karataliu commented on pkg/cloudprovider/providers/azure/azure_standard.go:
```diff
@@ -623,6 +630,22 @@ func (as *availabilitySet) ensureHostInPool(serviceName string, nodeName types.N
 	if !foundPool {
+		if as.useStandardLoadBalancer() && len(newBackendPools) > 0 {
+			// Although standard load balancer supports backends from multiple availability
+			// sets, the same interface couldn't reference more than one load balancer of
```
better to use full name  'interface' -> 'network interface', i got a bit confused when looking at the comment.

feiskyer replied:
so are you referring go interface? that make sense

feiskyer committed:
Add support of Azure standard load balancer and public IP

feiskyer committed:
some updates

karataliu commented:
/lgtm
